File: common/src/main/java/generations/gg/generations/core/generationscore/world/loot/LootPoolEntryTypes.java
package generations.gg.generations.core.generationscore.world.loot;

import dev.architectury.registry.registries.DeferredRegister;
import dev.architectury.registry.registries.RegistrySupplier;
import generations.gg.generations.core.generationscore.GenerationsCore;
import net.minecraft.core.registries.Registries;
import net.minecraft.world.level.storage.loot.Serializer;
import net.minecraft.world.level.storage.loot.entries.LootPoolEntryContainer;
import net.minecraft.world.level.storage.loot.entries.LootPoolEntryType;

import java.util.function.Supplier;

public class LootPoolEntryTypes {
    public static final DeferredRegister<LootPoolEntryType> REGISTER = DeferredRegister.create(GenerationsCore.MOD_ID, Registries.LOOT_POOL_ENTRY_TYPE);

    public static <T extends LootPoolEntryContainer> RegistrySupplier<LootPoolEntryType> register(String name, Supplier<Serializer<T>> supplier) {
        return REGISTER.register(name, () -> new LootPoolEntryType(supplier.get()));
    }

    public static final RegistrySupplier<LootPoolEntryType> RESOUCE_KEY = register("resource_key", ResouceKeyEntry.Serializer::new);

    public static void init() {
        REGISTER.register();
    }

}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/loot/LootPoolEntryTypes.kt
package generations.gg.generations.core.generationscore.common.world.loot

import com.mojang.serialization.MapCodec
import generations.gg.generations.core.generationscore.common.util.PlatformRegistry
import net.minecraft.core.Holder
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.world.level.storage.loot.entries.LootPoolEntryContainer
import net.minecraft.world.level.storage.loot.entries.LootPoolEntryType

object LootPoolEntryTypes: PlatformRegistry<LootPoolEntryType>(Registries.LOOT_POOL_ENTRY_TYPE, BuiltInRegistries.LOOT_POOL_ENTRY_TYPE) {
    val RESOURCE_KEY = register("resource_key", ResouceKeyEntry.CODEC)

    fun <T : LootPoolEntryContainer> register(name: String, codec: MapCodec<T>): Holder<LootPoolEntryType> {
        return create(name, { LootPoolEntryType(codec) })
    }
}
